### PR TITLE
Add Confusion Matrix.

### PR DESCRIPTION
**Pull request by df-um**

### CODE REVIEW

File: terms/confusionmatrix_df.tex [hunk 1]
```diff
@@ -0,0 +1,199 @@
+\chapter{Confusion Matrix}
+\label{ch:confusion-matrix}
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: Need indexing across different pages.  See how its done in the template.

File: terms/confusionmatrix_df.tex [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+A \textit{Confusion Matrix}\footnote{Also called a \textit{Contingency Table}.} (CM) is a cross-tabulation illustrating how well a model classifies categorical data. By convention \citep{sammut2017encyclopedia}, the CM of an N-class model is an N$\times$N matrix indexed by the true class in the row dimension and the predicted class in the column dimension (Table \ref{tab:cm_spam}).
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: Put in brackets (not material for a footnote).

File: ics5110_ml_handbook.tex [hunk 3]
```diff
@@ -221,6 +226,7 @@ \chapter{Introduction}
 %% the term and initials of the student in the filename.
 %% \input{terms/myterm_jpe.tex}
 \input{terms/crossvalidation_lmd.tex} %TODO
+\input{terms/confusionmatrix_df.tex}
```
Comment (jp-um) [inline ics5110_ml_handbook.tex]: In alphabetical order please ...

File: terms/confusionmatrix_df.tex [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    \textit{M\&M's}   & 34              & 3                 & 8  \\
+    \textit{Skittles} & 1               & 28                & 5  \\
+    \textit{Smarties} & 2               & 4                 & 22 \\
+    \bottomrule
+  \end{tabular}
+  \caption{CM of a hypothetical sweets classifier. The main diagonal of the CM shows the number of correct predictions, whilst the remaining elements indicate how many sweets were misclassified.}
+  \label{tab:cm_sweets}
+\end{table}
+
+The CM of the model $h : X \mapsto C$ over the concept $c : X \mapsto C$ using dataset $S \subset X$ is formally defined \citep{cichosz2014data} as a matrix $\Xi$ such that $\Xi_{c,S}(h)[d_1,d_2] = |S_{h=d_1,c=d_2}|$. The CM is constructed by incrementing the element corresponding to the true class \textit{vis-a-vi} the predicted class for each object in the dataset (Algorithm \ref{alg:cm}).
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: `\citep{cichosz2014data}` should be at end of sentence

File: terms/confusionmatrix_df.tex [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      \State $\Xi_{d_1,d_2} \gets \Xi_{d_1,d_2} + 1$
+    \EndFor
+  \end{algorithmic}
+  \caption{The CM is initialised to the zero matrix, and populated by iterating over all the objects $x$ with corresponding true class $d_1$ and predicted class $d_2$ and incrementing the element $(d_1,d_2)$ by 1 for each matching outcome.}
+  \label{alg:cm}
+\end{algorithm}
+
+In binary classification, the CM consists of 2 specially designated classes called the \textit{positive} class and the \textit{negative} class \citep{saito2015precision}. As indicated in Table \ref{tab:cm_binary}, positive outcomes from the true class  which are classified correctly are called \textit{True Positives} (TP), whilst misclassifications are called \textit{False Negatives} (FN). On the other hand, negative true class outcomes which are classified correctly are called \textit{True Negatives} (TN), and misclassifications are called \textit{False Positives} (FP). In natural sciences, FP are called \textit{Type I Errors} and FN are known as \textit{Type II Errors} \citep{fielding1997review}.
+
+\begin{table}[ht]
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: In general put some vertical space between tables (or equations) and the rest of the text.

File: terms/confusionmatrix_df.tex [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    \midrule
+    \textit{+ve}  & TP           & FN \\
+    \textit{-ve}  & FP           & TN \\
+    \bottomrule
+  \end{tabular}
+  \caption{CMs of binary classifiers have positive (+ve) and negative (-ve) classes, and elements called \textit{True Positives} (TP), \textit{False Positives} (FP), \textit{True Negatives} (TN) and \textit{False Negatives} (FN).}
+  \label{tab:cm_binary}
+\end{table}
+
+The information presented in the CM can be used to evaluate the performance of different binary classifiers \citep{lu2004predicting}. A number of statistics (Eq. \ref{eq:cm_acc}-\ref{eq:cm_fscore}) derived from the CM have been proposed in the literature \citep{deng2016improved} to gain a better understanding of what are the strengths and weaknesses of different classifiers. Caution should be exercised when interpreting metrics \citep{jeni2013facing}, since the CM could be misleading if the data is imbalanced and an important subrange of the domain (e.g. a rare diseases) is underrepresented \citep{raeder2012learning}.
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: use `e.g.\ ` so the space is not considered as one after the sentence.  also give an example of how imbalanced datasets can produce misleading accuracy

File: terms/confusionmatrix_df.tex [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+These metrics are important in situations in which a particular type of misclassification, i.e. FP or FN, could have worse consequences than the other \citep{hassanien2017advances}. For example, FP are more tolerable than FN in classifiers which predict whether a patient has a disease or not. Even though both misclassifications are undesirable, in medical applications it's better to err on the side of caution since FN could be fatal.
+
+\textit{Accuracy} (ACC) is the proportion of correct predictions (Eq. \ref{eq:cm_acc}). It is a class-insensitive metric because it can give a high rating to a model which classifies majority class objects correctly but misclassifies interesting minority class objects \citep{branco2016survey}. The other metrics should be preferred since they're more class-sensitive and give better indicators when the dataset is imbalanced.
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: Abbreviations, e.g. they're are for emails, not for scientific texts.

File: terms/confusionmatrix_df.tex [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+\end{equation}
+
+\textit{Positive Predictive Value} (PPV), or \textit{Precision}, is the ratio of the correct positive predictions from the total positive predictions (Eq. \ref{eq:cm_ppv}).
+
+\begin{equation}
+\label{eq:cm_ppv}
+PPV = \frac{|TP|}{|TP \cup FP|}
+\end{equation}
+
+Precision and Recall are borrowed from the discipline of \textit{Information Extraction} \citep{sokolova2009systematic}. A composite metric called \textit{F-score}, \textit{F1-score}, or \textit{F-measure} (Eq. \ref{eq:cm_fscore}) can be derived by finding their harmonic mean \citep{kelleher2015fundamentals}.
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: There is a nice image of a dart board explaining precision and recall ... please add.

File: terms/confusionmatrix_df.tex [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+\begin{equation}
+\label{eq:cm_fscore}
+\textnormal{\textit{F-score}} = 2 \times \frac{PPV \times TPR}{PPV + TPR}
+\end{equation}
+
+The complements of ACC, NPV, TNR, TPR and PPV are called, respectively, \textit{Error Rate}, \textit{False Omission Rate}, \textit{False Positive Rate}, \textit{False Negative Rate} and \textit{False Discovery Rate}.
+
+Using all metrics could be counterproductive due to information redundancy, but none of the metrics is enough on its own \citep{ma2007adequate}. For instance, Recall is class-sensitive but it would give a perfect score to an inept model which simply returns the positive class. Thus, the best approach is to evaluate with complementary pairs \citep{gu2009evaluation} such as Sensitivity \textit{vs.} Specificity, or Precision \textit{vs.} Recall; or a combined measure such as the F-score.
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: vs does not have a following period (like dr) ...

File: terms/confusionmatrix_df.tex [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    M\&M's    & Skittles $\cup$ Smarties \\
+    Skittles  & M\&M's $\cup$ Smarties \\
+    Smarties  & Skittles $\cup$ M\&M's \\
+    \bottomrule
+  \end{tabular}
+  \caption{2-class CMs derived through decomposition of the 3-class CM from Table \ref{tab:cm_sweets} using the 1-vs-rest approach.}
+  \label{tab:cm_1vsN}
+\end{table}
+
+Taking into account the above, CMs are suitable for visualising, evaluating, and comparing the performance of binary or multi-class classifiers. They should be used in conjunction with metrics such as the F-measure to avoid bias, especially if the dataset is unbalanced. For further details on the theoretical aspects of CMs and for practical examples in R refer to \citep{cichosz2014data}; for examples in Python refer to \citep{muller2016introduction}.
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: Excellent!

File: terms/confusionmatrix_df.tex [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    F-score      & 0.76   & 0.82 \\
+    \bottomrule
+  \end{tabular}
+  \caption{Statistics derived from the CMs in Fig. \ref{fig:cm_wines}.}
+  \label{tab:cm_metrics}
+\end{margintable}
+
+As it can be intuitively deduced from Fig. \ref{fig:cm_model}, the decision boundary of the non-linear model is a better fit than the linear model. The CMs in Fig. \ref{fig:cm_wines} also show that non-linear model performs better with a higher TP, and consequently lower TN. The biggest advantage of the non-linear model is the higher Sensitivity resulting in a better F-score.
+
+\begin{figure}
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: if you lift images from somewhere else you need to specify that in the caption (Taken from [Ebejer et al.])

File: terms/confusionmatrix_df.tex [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  \includegraphics{confusion_matrix/cm.png}
+  \caption{The linear classifier has 16 TP, 10 TN, 7 FN and 3 FP, whilst the non-linear classifier has 18 TP, 10 TN, 5 FN and 3 FP.}
+  \label{fig:cm_wines}
+\end{figure}
+
+\index{class options|)}
```
Comment (jp-um) [inline terms/confusionmatrix_df.tex]: index class options ???? no .... this should be a closing \index{confusion matrix|)} but you need an opening tag.

No indexing information throughout ... please add.

File: terms/confusionmatrix_df.tex (hunk 8 above)
Comment (df-um) [reply terms/confusionmatrix_df.tex]: The ones I found (and ended up adding to the doc) are about the difference between accuracy and precision.

**Comment by df-um** [comment]
I should have addressed all comments. Please advise if more changes are necessary.

**Comment by df-um** [comment]
I fixed the vis-a-vis typo and remove keywords not directly related to CMs from the index.

**Comment by jp-um** [comment]
Merged.

**Pull request closed without merging**